### PR TITLE
feat: add mock vertx backend for dev purposes

### PR DESCRIPTION
**Pull request by matthew-chirgwin**
## Status
**READY**

### Commit Message Title
* When opening this PR, the raiser should set the title of this PR to the first line of their desired commit message, e.g:
```
feat: add mock vertx backend for dev purposes
```
* The reviewer should ensure that the first commit message field is of this form when performing the `squash and merge` from this page.

### Commit Message Description
```
- adds a mock backend which can be used for dev/test purposes

Closes: #22

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>
```
* The reviewer should copy the above text into the extended description field when performing the `squash and merge` from this page.

## Checklist
- [ ] Automated tests exist
- [X] Documentation exists
- [ ] Local unit tests performed
- [X] Sufficient logging/trace
- [X] Desired commit message set as PR title and commit description set above

### CODE REVIEW

**Comment by matthew-chirgwin** [comment]
For completeness - this is a little JS function I put together to drive this from a browser:

```
const runSocket = (time = 5000, endpoint = 'ws://localhost:8080/demoproduce') => {
    const socket = new WebSocket(endpoint);
    const socketRoot = endpoint.substr(endpoint.lastIndexOf('/') + 1, endpoint.length);
    const log = (msg, content) => {
        console.log(`${socketRoot} - ${new Date()} - ${msg}`);
        content ? console.dir(content) : false;
    };

    socket.addEventListener('open', event => {
        log('Open', event);
        socket.send('{"action":"start"}');
        setTimeout(() => {
            socket.send('{"action":"stop"}');
            socket.close();
            log('Closing');
        }, time);
    });

    socket.addEventListener('error', event => {
        log('Error message from server', event);
    });

    socket.addEventListener('message', event => {
        log('Message from server', event);
    });

    socket.addEventListener('close', event => {
        log('Socket closed', event);
    });
};

const runEmulation = (port = 8080) => {
    runSocket(30000, `ws://localhost:${port}/demoproduce`);
    runSocket(30000, `ws://localhost:${port}/democonsume`);
};
```

File: ui/src/DevUtils/MockVertx/MockVertxServer.js [hunk 1]
```diff
@@ -0,0 +1,292 @@
+const ws = require('ws');
+const {
+  generateMockSuccessShape,
+  generateMockConsumerMetadataResponseShape,
+  generateMockProducerMetadataResponseShape,
+  generateMockErrorResponseShape,
+} = require('../../TestUtils/server.testutil.js');
```
[the remark this answers is not in the document]
Comment (matthew-chirgwin) [reply ui/src/DevUtils/MockVertx/MockVertxServer.js:7]: As run in node, the config to Jest/Webpack dont apply - if it is possible to change this I am more than happy to :)

[the remark this answers is not in the document]
Comment (JQrdan) [reply ui/src/DevUtils/MockVertx/MockVertxServer.js:7]: Ahhh yes of course!

File: ui/src/DevUtils/MockVertx/MockVertxServer.js [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+  };
+
+  const handleProduceMessage = (
+    { action, custom },
+    ws,
+    intervalTimer = produceRate
+  ) => {
+    // if action start, kick off an interval to add to the to be consumed array
+    // if stop, cancel that timer
+    if (action === 'start' && !mockProducerInterval) {
```
Comment (nictownsend) [inline ui/src/DevUtils/MockVertx/MockVertxServer.js:112]: Clarification question - double checking new sample app sends one message on websocket to "start producing with payload x"?

Comment (matthew-chirgwin) [reply ui/src/DevUtils/MockVertx/MockVertxServer.js:112]: Client sends a payload ala: `{action: 'start', payload: 'something'}`. If no payload, the backend sends `Hello world`

File: ui/src/DevUtils/MockVertx/MockVertxServer.js [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+const generateModelForConfig = ({
+  produceRate = TICK_RATE,
+  topic = TOPIC_NAME,
+  numberOfPartitions = NUMBER_OF_PARTITIONS,
+  defaultPayload = PLACEHOLDER_PAYLOAD,
+  maxOffsets = MAX_OFFSETS,
+  enableLog = false,
+  errorChance = ERROR_CHANCE,
+}) => {
```
[the remark this answers is not in the document]
Comment (matthew-chirgwin) [reply ui/src/DevUtils/MockVertx/MockVertxServer.js:55]: Good spot! Fixed

File: ui/src/DevUtils/MockVertx/MockVertxServer.js [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+const PLACEHOLDER_PAYLOAD = 'Hello world';
+const TICK_RATE = 2000;
+const ERROR_CHANCE = 5;
+
+// Private helper functions
+
+// maps an object of event/handler pairs (config) to the websocket server
+const handleWebsocketEvents = (config = {}) => (ws, { url }) => {
+  Object.entries(config).forEach(([event, handler]) => {
+    ws.on(event, (evt) =>
```
[the remark this answers is not in the document]
Comment (matthew-chirgwin) [reply ui/src/DevUtils/MockVertx/MockVertxServer.js:26]: Good idea - updated

**Pull request merged**